File: .pub-cache/hosted/pub.dartlang.org/flutter_socket_io-0.6.0/example/ios/Runner/SocketObj/SocketIO.m
//
//  SocketIO.m
//  SocketTest
//
//  Created by AnhNguyen on 8/18/18.
//  Copyright © 2018 ATA_Studio. All rights reserved.
//

#import "SocketIO.h"
#import "UtilsSocket.h"
#import "SocketListener.h"

/*--------------------------------------------------------------------
 * Macro weak and strong self
 ---------------------------------------------------------------------*/
#define weakify(object) __weak typeof(object)weakSelf = object
#define strongify(object) __strong typeof(object)strongSelf = object

#define EVENT_CONNECT               @"connect"
#define EVENT_CONNECTING            @"connecting"
#define EVENT_DISCONNECT            @"disconnect"
#define EVENT_ERROR                 @"error"
#define EVENT_MESSAGE               @"message"
#define EVENT_CONNECT_ERROR         @"connect_error"
#define EVENT_CONNECT_TIMEOUT       @"connect_timeout"
#define EVENT_RECONNECT             @"reconnect"
#define EVENT_RECONNECT_ERROR       @"reconnect_error"
#define EVENT_RECONNECT_FAILED      @"reconnect_failed"
#define EVENT_RECONNECT_ATTEMPT     @"reconnect_attempt"
#define EVENT_RECONNECTING          @"reconnecting"
#define EVENT_PING                  @"ping"
#define EVENT_PONG                  @"pong"

@interface SocketIO()
@end

@implementation SocketIO

+ (instancetype)initSocketIO:(FlutterMethodChannel *)methodChannel
                       query:(NSDictionary *)query
                      domain:(NSString *)domain
                   nameSpace:(NSString *)nameSpace
              statusCallback:(NSString *)callBack {
    SocketIO *obj = [[SocketIO alloc] init];
    obj.methodChannel = methodChannel;
    obj.domain = domain;
    obj.query = query;
    obj.nameSpace = nameSpace;
    obj.statusCallback = callBack;
    obj.subscribes = [[NSMutableDictionary alloc] init];
    return obj;
}

- (void)removeChannelAll {
    if (self.subscribes) {
        [self.subscribes removeAllObjects];
    }
    self.subscribes = nil;
}

- (NSString *)getId {
    return [self getSocketUrl];
}

- (NSString *)getSocketUrl {
    if (self.nameSpace) {
        return [NSString stringWithFormat:@"%@%@", self.domain, self.nameSpace];
    } else {
        return self.domain;
    }
}

- (void)initSocket {
    if (self.socket) {
        if ([self isConnected]) {
            [self disconnect];
        }
        self.socket = nil;
    }
    @try {
        NSURL *url = [NSURL URLWithString:self.domain];
        if (!manager) {
            manager = [[SocketManager alloc] initWithSocketURL:url config:@{@"log": @YES, @"compress": @YES, @"connectParams": self.query, @"forceWebsockets": @YES }];
        }
        self.socket = manager.defaultSocket;
        self.socket = [manager socketForNamespace:self.nameSpace];
        weakify(self);
        [self.socket on:EVENT_CONNECT callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_CONNECT obj:ack];
        }];
        
        [self.socket on:EVENT_RECONNECT callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_RECONNECT obj:ack];
        }];
        
        [self.socket on:EVENT_RECONNECTING callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_RECONNECTING obj:ack];
        }];
        
        [self.socket on:EVENT_RECONNECT_ATTEMPT callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_RECONNECT_ATTEMPT obj:ack];
        }];
        
        [self.socket on:EVENT_RECONNECT_FAILED callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_RECONNECT_FAILED obj:ack];
        }];
        
        [self.socket on:EVENT_RECONNECT_ERROR callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_RECONNECT_ERROR obj:ack];
        }];
        
        [self.socket on:EVENT_CONNECT_TIMEOUT callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_CONNECT_TIMEOUT obj:ack];
        }];
        
        [self.socket on:EVENT_DISCONNECT callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_DISCONNECT obj:ack];
        }];
        
        [self.socket on:EVENT_CONNECT_ERROR callback:^(NSArray* data, SocketAckEmitter* ack) {
            strongify(weakSelf);
            [strongSelf onSocketCallback:EVENT_CONNECT_ERROR obj:ack];
        }];
        
        [self.socket on:@"socket_info" callback:^(NSArray* data, SocketAckEmitter* ack) {
            NSLog(@"aa");
        }];

    } @catch (NSException *exception) {
        NSLog(@"CONNECT FAIL : %@", exception.description);
        if (self.methodChannel && self.statusCallback) {
            [self onSocketCallback:self.statusCallback obj:@"failed"];
        }
    }
    
}

- (void)connect {
    if (!self.socket) {
        NSLog(@"SOCKET %@ IS NOT INITIALIZED", [self getId]);
        return;
    }
    if ([self isConnected]) {
        NSLog(@"SOCKET %@ IS ALREADY CONNECTED", [self getId]);
        return;
    }
    NSLog(@"CONNECTING SOCKET %@", [self getId]);
    [self.socket connect];
    
}

- (void)sendMessage:(NSString *)eventName mess:(NSString *)message callBack:(NSString *)callBack {
    if ([self isConnected] && eventName && message) {
        NSLog(@"SEND MESSAGE EVENT %@ MESSAGE %@", eventName, message);
        SocketListener *listener = [SocketListener initSocketListener:self.methodChannel event:eventName socketId:[self getId] callBack:callBack];
        NSArray *jsonObject = [NSJSONSerialization JSONObjectWithData:[message dataUsingEncoding:NSUTF8StringEncoding]
                                                              options:0 error:NULL];
        [[self.socket emitWithAck:eventName with:@[jsonObject]] timingOutAfter:0 callback:^(NSArray* data) {
            [listener call:data];
        }];
    } else {
        NSLog(@"INVALID PARAMS : event or message is NULL or EMPTY!");
    }
}

- (void)subscribe:(NSString *)eventName callBack:(NSString *)callBack {
    if (self.socket && manager && eventName) {
        NSMutableArray *listeners = [self.subscribes objectForKey:eventName];
        if (!listeners) {
            listeners = [[NSMutableArray alloc] init];
        }
        
        SocketListener *listener = [SocketListener initSocketListener:self.methodChannel event:eventName socketId:[self getId] callBack:callBack];
        if (callBack && [UtilsSocket isExistedCallback:listeners callBack:callBack]) {
            [listeners addObject:listener];
        }
        
        [self.subscribes setValue:listeners forKey:eventName];
        [self.socket on:eventName callback:^(NSArray* data, SocketAckEmitter* ack) {
            [listener call:data];
        }];
    }
}

- (void)subscribeList:(NSMutableDictionary *)sub {
    if (self.socket && manager && ![UtilsSocket isNullOrEmpty:sub]) {
        NSArray *keys = [sub allKeys];
        for (id key in keys) {
            [self subscribe:key callBack:[sub objectForKey:key]];
        }
    }
}

- (void)unSubscribe:(NSString *)eventName callBack:(NSString *)callBack {
    if (self.socket && manager && ![UtilsSocket isNullOrEmpty:self.subscribes]) {
        NSMutableArray *listeners = [self.subscribes objectForKey:eventName];
        if ([UtilsSocket isNullOrEmptyArray:listeners]) {
            [self.subscribes removeObjectForKey:eventName];
            [self.socket off:eventName];
        } else {
            SocketListener *listener = [UtilsSocket findListener:listeners callBack:callBack];
            if (listener) {
                [listeners removeObject:listener];
                if (listeners.count < 1) {
                    [self.subscribes removeObjectForKey:eventName];
                    [self.socket off:eventName];
                } else {
                    [self.subscribes setValue:listeners forKey:eventName];
                    [self.socket off:eventName];
                }
            }
        }
    }
}

- (void)unSubscribeList:(NSMutableDictionary *)sub {
    if (self.socket && manager && ![UtilsSocket isNullOrEmpty:sub]) {
        NSArray *keys = [sub allKeys];
        for (id key in keys) {
            [self unSubscribe:key callBack:[sub objectForKey:key]];
        }
    }
}

- (void)unSubscribeAll {
    if (self.socket && manager && ![UtilsSocket isNullOrEmpty:self.subscribes]) {
        NSArray *keys = [self.subscribes allKeys];
        for (id key in keys) {
            [self unSubscribe:key callBack:nil];
        }
    }
}

- (BOOL)isConnected {
    if (self.socket && manager) {
        if (self.socket.status == SocketIOStatusConnected) {
            return YES;
        }
    }
    return NO;
}

- (void)disconnect {
    if (self.socket && manager) {
        [self.socket disconnect];
        [manager disconnect];
    }
}

- (void)onSocketCallback:(NSString *)status obj:(id)arg {
    if (self.methodChannel && self.statusCallback) {
        [self.methodChannel invokeMethod:[NSString stringWithFormat:@"%@|%@|%@", [self getId], self.statusCallback, self.statusCallback] arguments:status];
    }
    if (arg) {
        NSLog(@"onSocketCallback %@", self.statusCallback);
    }
}

- (void)destroy {
    [self disconnect];
    [self unSubscribeAll];
    [self removeChannelAll];
    self.socket = nil;
    manager = nil;
    self.methodChannel = nil;
    self.nameSpace = nil;
    self.domain = nil;
    self.statusCallback = nil;
    [self.subscribes removeAllObjects];
    self.subscribes = nil;
}

@end

File: .pub-cache/hosted/pub.dartlang.org/flutter_socket_io-0.6.0/android/src/main/java/com/itsclicking/clickapp/fluttersocketio/SocketIO.java
package com.itsclicking.clickapp.fluttersocketio;

import android.os.Handler;
import android.os.Looper;

import com.google.gson.Gson;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ConcurrentMap;

import io.flutter.plugin.common.MethodChannel;
import io.socket.client.Ack;
import io.socket.client.IO;
import io.socket.client.Manager;
import io.socket.client.Socket;
import io.socket.client.Url;
import io.socket.emitter.Emitter;
import io.socket.engineio.client.transports.WebSocket;

public class SocketIO {

    private static final String TAG = "SocketIO";

    private MethodChannel _methodChannel;
    private String _domain;
    private String _namespace;
    private String _statusCallback;
    private String _query;
    private Socket _socket;
    private IO.Options mOptions;
    private ConcurrentMap<String, ConcurrentLinkedQueue<SocketListener>> _subscribes;
    private static final ConcurrentHashMap<String, Manager> managers = new ConcurrentHashMap<String, Manager>();

    public SocketIO(MethodChannel methodChannel, String domain,
                    String namespace, String query, String statusCallback) {
        _methodChannel = methodChannel;
        _domain = domain;
        _namespace = namespace;
        _query = query;
        _statusCallback = statusCallback;
        _subscribes = new ConcurrentHashMap<>();
    }

    private void removeChannelAll() {
        if (_subscribes != null) {
            _subscribes.clear();
        }
    }

    private String getSocketUrl() {
        return _domain + (_namespace == null ? "" : _namespace);
    }

    private void dumpChannelsCount() {
        if (!Utils.isNullOrEmpty(_subscribes)) {
            Utils.log("socketInfo", "SUBSCRIBES SIZES: " + _subscribes.size());
            for (Map.Entry<String, ConcurrentLinkedQueue<SocketListener>> item : _subscribes.entrySet()) {
                ConcurrentLinkedQueue<SocketListener> listeners = item.getValue();
                if (listeners == null) {
                    Utils.log("socketInfo", "CHANNEL: " + item.getKey() + " with TOTAL LISTENERS: NULL");
                } else {
                    Utils.log("socketInfo", "CHANNEL: " + item.getKey() + " with TOTAL LISTENERS: " + listeners.size());
                }
            }
        } else {
            Utils.log("socketInfo", "SUBSCRIBES SIZES: NULL or EMPTY");
        }
    }

    private Socket getSocket() {
        URI source;
        URL parsed;

        try {
            URI uri = new URI(getSocketUrl());
            parsed = Url.parse(uri);
            source = parsed.toURI();
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
        }

        mOptions = new IO.Options();
        mOptions.transports = new String[]{WebSocket.NAME};

        if (!Utils.isNullOrEmpty(_query)) {
            Utils.log(TAG, "query: " + _query);
            mOptions.query = _query;
        }

        String id = Url.extractId(parsed);
        boolean newConnection = !managers.containsKey(id);
        Manager io;

        if (newConnection) {
            io = new Manager(source, mOptions);
            managers.putIfAbsent(id, io);
        } else {
            if (!managers.containsKey(id)) {
                managers.putIfAbsent(id, new Manager(source, mOptions));
            }
            io = managers.get(id);
        }

        return io.socket(_namespace, mOptions);
    }

    public String getId() {
        return getSocketUrl();
    }

    private void onSocketCallback(final String status, Object... args) {
        if (_methodChannel != null && !Utils.isNullOrEmpty(_statusCallback)) {
            final Handler _handler = new Handler(Looper.getMainLooper());
            _handler.post(new Runnable() {
                @Override
                public void run() {
                    _methodChannel.invokeMethod(getId() + "|" + _statusCallback + "|" + _statusCallback, status);
                }
            });
        }
        if (args != null) {
            Utils.log(TAG, status + ": " + new Gson().toJson(args));
        }
    }

    public void init() {
        if (_socket != null) {
            if (_socket.connected()) {
                _socket.disconnect();
            }
            _socket = null;
        }

        _socket = getSocket();

        Utils.log(TAG, "connecting..." + _socket.id());

        //start listen connection events
        _socket.on(Socket.EVENT_CONNECT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_CONNECT, args);
            }
        }).on(Socket.EVENT_RECONNECT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_RECONNECT, args);
            }
        }).on(Socket.EVENT_RECONNECTING, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_RECONNECTING, args);
            }
        }).on(Socket.EVENT_RECONNECT_ATTEMPT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_RECONNECT_ATTEMPT, args);
            }
        }).on(Socket.EVENT_RECONNECT_FAILED, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_RECONNECT_FAILED, args);
            }
        }).on(Socket.EVENT_RECONNECT_ERROR, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_RECONNECT_ERROR, args);
            }
        }).on(Socket.EVENT_CONNECT_TIMEOUT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_CONNECT_TIMEOUT, args);
            }
        }).on(Socket.EVENT_DISCONNECT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_DISCONNECT, args);
            }
        }).on(Socket.EVENT_CONNECT_ERROR, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                onSocketCallback(Socket.EVENT_CONNECT_ERROR, args);
            }
        });
        //end listen connection events
    }

    public void connect() {
        if (_socket == null) {
            Utils.log(TAG, "socket: " + getId() + " is not initialized!");
            return;
        }
        if (_socket.connected()) {
            Utils.log(TAG, "socket: " + getId() + " is already connected");
            return;
        }
        Utils.log(TAG, "connecting socket: " + getId());
        _socket.connect();
    }

    public void sendMessage(String event, String message, final String callback) {

        if (Utils.isNullOrEmpty(event) || Utils.isNullOrEmpty(message)) {
            Utils.log("sendMessage", "Invalid params: event or message is NULL or EMPTY!");
        } else if (isConnected()) {
            Utils.log("sendMessage", "Event: " + event + " - with message: " + message);
            JSONObject jb = null;
            try {
                jb = new JSONObject(message);
            } catch (JSONException e) {
                e.printStackTrace();
            }
            if (jb != null) {
                final SocketListener listener = new SocketListener(_methodChannel, getId(), event, callback);
                _socket.emit(event, jb, new Ack() {
                    @Override
                    public void call(Object... args) {
                        listener.call(args);
                    }
                });
            }
        }

    }

    public void subscribe(String event, final String callback) {
        Utils.log("subscribe", "channel: " + event + " - with callback: " + callback);
        if (Utils.isNullOrEmpty(event)) {
            Utils.log("subscribe", "Invalid params: event is NULL/EMPTY!");
        } else {
            if (_socket != null) {
                dumpChannelsCount();
                ConcurrentLinkedQueue<SocketListener> listeners = _subscribes.get(event);

                if (listeners == null) {
                    listeners = new ConcurrentLinkedQueue<>();
                }

                SocketListener listener = new SocketListener(_methodChannel, getId(), event, callback);

                if (!Utils.isNullOrEmpty(callback) && !Utils.isExisted(listeners, callback)) {
                    listeners.add(listener);
                }

                _subscribes.put(event, listeners);
                _socket.on(event, listener);

                dumpChannelsCount();
            } else {
                Utils.log("subscribe", "socket is NULL");
            }
        }
    }

    public void subscribes(Map<String, String> subscribes) {
        if (Utils.isNullOrEmpty(subscribes)) {
            Utils.log("subscribes", "Subscribes list is NULL or EMPTY!");
        } else if (_socket != null) {
            Utils.log(TAG, "--- subscribes ---" + new Gson().toJson(subscribes));
            for (Map.Entry<String, String> sub : subscribes.entrySet()) {
                if (!Utils.isNullOrEmpty(sub.getKey())) {
                    subscribe(sub.getKey(), sub.getValue());
                }
            }
        }
    }

    public void unSubscribe(String eventName, final String callback) {
        Utils.log("unSubscribe", "channel: " + eventName + " - with callback: " + callback);
        if (Utils.isNullOrEmpty(eventName)) {
            Utils.log("unSubscribe", "Invalid params: event is NULL or EMPTY!");
        } else {
            if (_socket != null) {
                dumpChannelsCount();
                ConcurrentLinkedQueue<SocketListener> listeners = _subscribes.get(eventName);
                if (listeners == null || Utils.isNullOrEmpty(callback)) {
                    _subscribes.remove(eventName);
                    _socket.off(eventName);
                } else {
                    SocketListener listener = Utils.findListener(listeners, callback);
                    if (listener != null) {
                        listeners.remove(listener);
                        if (listeners.size() < 1) {
                            _subscribes.remove(eventName);
                            _socket.off(eventName);
                        } else {
                            _subscribes.put(eventName, listeners);
                            _socket.off(eventName, listener);
                        }
                    }
                }
                dumpChannelsCount();
            } else {
                Utils.log("unSubscribe", "socket is NULL");
            }
        }
    }

    public void unSubscribes(Map<String, String> unSubscribes) {
        if (Utils.isNullOrEmpty(unSubscribes)) {
            Utils.log("unSubscribes", "unSubscribes list is NULL or EMPTY!");
        } else if (_socket != null) {
            Utils.log(TAG, "--- unSubscribes ---" + new Gson().toJson(unSubscribes));
            for (Map.Entry<String, String> sub : unSubscribes.entrySet()) {
                if (!Utils.isNullOrEmpty(sub.getKey())) {
                    unSubscribe(sub.getKey(), sub.getValue());
                }
            }
        }
    }

    public void unSubscribesAll() {
        if (_socket != null && !Utils.isNullOrEmpty(_subscribes)) {
            for (Map.Entry<String, ConcurrentLinkedQueue<SocketListener>> sub : _subscribes.entrySet()) {
                if (!Utils.isNullOrEmpty(sub.getKey())) {
                    unSubscribe(sub.getKey(), null);
                }
            }
        }
    }

    public boolean isConnected() {
        if (_socket != null) {
            Utils.log(TAG, "socket id: " + getId() + " is connected: " + _socket.connected());
            return _socket.connected();
        } else {
            Utils.log(TAG, "socket id: " + getId() + " is NULL");
        }
        return false;
    }

    public void disconnect() {
        if (_socket != null) {
            _socket.disconnect();
        }
    }

    public void destroy() {
        Utils.log(TAG, "--- START destroy ---");
        disconnect();
        unSubscribesAll();
        removeChannelAll();
        _socket = null;
        _namespace = null;
        _statusCallback = null;
        mOptions = null;
        _methodChannel = null;
        _subscribes = null;
        managers.clear();
        Utils.log(TAG, "--- END destroy ---");
    }
}
